src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborWriter.Map.cs

```diff
@@ -38,14 +41,135 @@ public void WriteEndMap()
+
+            if (CborConformanceLevelHelpers.RequiresUniqueKeys(ConformanceLevel))
+            {
+                SortedSet<(int offset, int keyLength, int keyValueLength)> ranges = GetKeyValueEncodingRanges();
+
+                (int offset, int keyLength, int valueLength) currentKeyRange =
+                    (_currentKeyOffset.Value,
+                     _currentValueOffset.Value - _currentKeyOffset.Value,
+                     0);
+
+                if (ranges.Contains(currentKeyRange))
+                {
+                    // TODO: check if rollback is necessary here
+                    throw new InvalidOperationException("Duplicate key encoding in CBOR map.");
+                }
+            }
+        }
+
+        private void HandleValueWritten()
+        {
+            Debug.Assert(_currentKeyOffset != null && _currentValueOffset != null);
+
+            if (CborConformanceLevelHelpers.RequiresUniqueKeys(ConformanceLevel) || CborConformanceLevelHelpers.RequiresSortedKeys(ConformanceLevel))
+            {
+                SortedSet<(int offset, int keyLength, int keyValueLength)> ranges = GetKeyValueEncodingRanges();
+
+                (int offset, int keyLength, int keyValueLength) currentKeyRange =
+                    (_currentKeyOffset.Value,
+                     _currentValueOffset.Value - _currentKeyOffset.Value,
+                     _offset - _currentKeyOffset.Value);
+
+                ranges.Add(currentKeyRange);
+            }
+
+            // reset state
+            _currentKeyOffset = _offset;
+            _currentValueOffset = null;
+        }
+
+        private void SortKeyValuePairEncodings()
+        {
+            if (_keyValueEncodingRanges == null)
+            {
+                return;
+            }
+
+            int totalMapPayloadEncodingLength = _offset - _frameOffset;
+            byte[] tempBuffer = s_bufferPool.Rent(totalMapPayloadEncodingLength);
+            Span<byte> tmpSpan = tempBuffer.AsSpan(0, totalMapPayloadEncodingLength);
+
+            // copy sorted ranges to temporary buffer
+            Span<byte> s = tmpSpan;
+            foreach((int offset, int keyLength, int valueLength) range in _keyValueEncodingRanges)
+            {
+                ReadOnlySpan<byte> kvEnc = GetKeyValueEncoding(range);
+                kvEnc.CopyTo(s);
+                s = s.Slice(kvEnc.Length);
+            }
+            Debug.Assert(s.IsEmpty);
+
+            // now copy back to the original buffer segment
+            tmpSpan.CopyTo(_buffer.AsSpan(_frameOffset, totalMapPayloadEncodingLength));
+
+            s_bufferPool.Return(tempBuffer, clearArray: true);
+        }
+
+        private ReadOnlySpan<byte> GetKeyEncoding((int offset, int keyLength, int valueLength) keyValueRange)
+        {
+            return _buffer.AsSpan(keyValueRange.offset, keyValueRange.keyLength);
+        }
+
+        private ReadOnlySpan<byte> GetKeyValueEncoding((int offset, int keyLength, int keyValueLength) keyValueRange)
```

Comment:
The (int,int,int) fields aren't named consistently.  On GetKeyEncoding it was valueLength (should be `ValueLength`), here it's **key**ValueLength (should be `KeyValueLength`).  Assuming they're the same thing, they should ideally share naming.